### DISCUSSION TITLE
Best examples for pagination in FastAPI?

### DISCUSSION THREAD

**Discussion by apiguy123** [fastapi/fastapi#9456] [Questions] [answered]
### First Check

- [ ] I added a very descriptive title here.
- [ ] I used the GitHub search to find a similar question and didn't find it.
- [ ] I searched the FastAPI documentation, with the integrated search.
- [ ] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [ ] I already read and followed all the tutorial in the docs and didn't find an answer.
- [ ] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [ ] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [ ] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [ ] I commit to help with one of those options 👆

### Example Code

```python
no code
```

### Description

Hi guys, I am new to API and FASTapi and have an abstract question.

How do you set up a process and models for pagination accessing SQL tables?

Is pagination set on the FastAPI side, or have you set it up in SQL models (SQLalchemy)?

I was thinking about two methods, but both are unsuitable.

Via FastAPI, you get a request, and the SQL response has many records. In FastAPI, you can declare data length and pagination system, but this would mean you need to repeatedly call for the same SQL data (which can be huge), which would be less efficient.
What if something changes in data while the client has already proceeded with pagination requests? This can cause inconsistency.
Could you point me to some solutions on doing pagination on FastAPI? A newbie would be happy for explanations and simple examples for better understanding.

Thanks a lot!

### Operating System

Windows

### FastAPI Version

0.95.0

### Python Version

3.11

**Comment by jgould22** [comment, marked as answer]
Generally when designing web apps you let the database handle pagination. 

There are a number of strategies you can use depending on the database in question.

Here is a blog post by citus that explains some of them using postgresql (it is pretty old but still generally relevant)

https://www.citusdata.com/blog/2016/03/30/five-ways-to-paginate/

**Comment by mateoradman** [comment]
Hi, I would suggest you to use SQLAlchemy for pagination. More specifically, use OFFSET and LIMIT SQL keywords to paginate records in your database. Here is a basic, incomplete example of how it looks like. You can take this example and transform it into what you're working with. I'd suggest creating Pydantic models for properly serializing the response, and returning the total count of all records along with the paginated list of Models. This way, the user of your API knows how many total records there are and can immediately calculate how many pages there are in total.

```python
from typing import Annotated
from sqlalchemy.orm import Session, sessionmaker
from sqlalchemy import Column, select
from sqlalchemy.sql import Executable
from fastapi import FastAPI, Depends

app = FastAPI()


# this is your sqlalchemy model
class Model:
    id: int = Column(...)
    # your sqlalchemy columns ...
    pass


# function that queries the db
def paginate(
    db: Session, offset: int | None = None, limit: int | None = None
) -> list[Model]:
    stmt: Executable = select(Model).offset(offset).limit(limit)
    result: list[Model] = db.execute(stmt).scalars().all()
    return result


# instance of a sessionmaker which, when instantiated, inits a new session
DBSession = sessionmaker()


# dependency which inits the db session
def get_db():
    db = DBSession()
    try:
        yield db
    finally:
        db.close()


@app.get("/items/")
def read_items(
    db: Annotated[Session, Depends(get_db)], offset: int = 0, limit: int = 10
):
    return paginate(db, offset, limit)


# calling the url http://127.0.0.1:8000/items/?offset=0&limit=10 will paginate first 10 records
```

**Comment by chrisK824** [comment]
I think that the easiest way going at this  https://uriyyo-fastapi-pagination.netlify.app/
